iree/compiler/InputConversion/MHLO/MHLOToLinalgOnTensors.cpp

```diff
@@ -215,6 +216,210 @@ struct FftOpConversion : public OpConversionPattern<mhlo::FftOp> {
+      return ShapedType::kDynamicSize;
+    return a * b;
+  }
+
+  LogicalResult matchAndRewrite(
+      mhlo::ScatterOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto operand = op.operand();
+    auto indices = op.scatter_indices();
+    auto updates = op.updates();
+
+    auto operandTy = operand.getType().dyn_cast<RankedTensorType>();
+    auto indicesTy = indices.getType().dyn_cast<RankedTensorType>();
+    auto updatesTy = updates.getType().dyn_cast<RankedTensorType>();
+    if (!operandTy || !indicesTy || !updatesTy) return failure();
+
+    if (indicesTy.getRank() < 2) return failure();
+
+    auto dimNumbers = op.scatter_dimension_numbers();
+    int64_t batchDims = dimNumbers.getIndexVectorDim();
+    int64_t numIndices = indicesTy.getDimSize(batchDims);
+
+    // No support for a dynamic number of indices.
+    if (batchDims != indicesTy.getRank() - 1 ||
+        indicesTy.isDynamicDim(batchDims))
+      return failure();
+
+    // Bail on dynamic update dimensions right now to avoid slice validation.
+    for (int i = batchDims, s = updatesTy.getRank(); i < s; i++) {
+      if (updatesTy.isDynamicDim(i)) return failure();
+    }
+
+    Location loc = op.getLoc();
+    ImplicitLocOpBuilder b(loc, rewriter);
+
+    // Make insertion dimensions explicit.
+    llvm::SmallVector<int64_t> updatesShape;
+    for (int i = 0; i < batchDims; i++)
+      updatesShape.push_back(updatesTy.getDimSize(i));
+    updatesShape.resize(operandTy.getRank() + batchDims, 1);
+    SmallVector<ReassociationIndices> map;
+    for (int i = 0; i < batchDims; i++) map.emplace_back(1, i);
+
+    auto updateWindowDims = dimNumbers.getUpdateWindowDims();
+    for (int operandDim = 0; operandDim < operandTy.getRank(); operandDim++) {
+      int64_t updatesDim = operandDim + batchDims;
+      auto begin = updateWindowDims.begin();
+      auto end = updateWindowDims.end();
+      if (std::find(begin, end, operandDim) == end) {
+        map.back().emplace_back(updatesDim);
+        continue;
+      }
+      updatesShape[updatesDim] = updatesTy.getDimSize(map.size());
+      map.emplace_back(1, updatesDim);
+    }
+
+    if (updatesTy.getRank() < updatesShape.size()) {
+      updatesTy =
+          RankedTensorType::get(updatesShape, updatesTy.getElementType());
+      updates = b.create<tensor::ExpandShapeOp>(updatesTy, updates, map);
+    }
+
+    // Collapse batch dimensions together so that we iterate over all batch
+    // dimensions together.
+    indices = collapseFrontDimsIfNeeded(indices, batchDims, b);
+    updates = collapseFrontDimsIfNeeded(updates, batchDims, b);
+    indicesTy = indices.getType().cast<RankedTensorType>();
+    updatesTy = updates.getType().cast<RankedTensorType>();
+
+    // Grab the first update dimension before it is merged with the batch. We
+    // already verified this must be static.
+    int64_t firstUpdateStaticDim = updatesTy.getDimSize(1);
```

Comment:
Why first update dimension? Is it because of the attribute `inserted_window_dims = [1]`?

Reply:
We fuse the batch dimensions with the first dimension. This allows us to iterate through slices of the update dims, slice, and insert into the operand. It avoids collapsing out the length-1 batch dim each iteration.